bpo-23395: Raise an exception if the SIGINT signal is ignored or not handled

mcepl opened the pull request:
Attached interrupt_main.patch fixes for ``_thread.interrupt_main()``. Raise an exception if the ``SIGINT`` signal is ignored (``SIG_IGN``) or not handled by Python (``SIG_DFL``).

The patch updates the documentation and adds unit tests.

<!-- issue-number: [bpo-23395](https://www.bugs.python.org/issue23395) -->
https://bugs.python.org/issue23395
<!-- /issue-number -->

vstinner committed:
Raise an exception if the SIGINT signal is ignored or not handled

Attached interrupt_main.patch fixes for ``_thread.interrupt_main()``.
Raise an exception if the ``SIGINT`` signal is ignored (``SIG_IGN``) or
not handled by Python (``SIG_DFL``).

The patch updates the documentation and adds unit tests.

mcepl committed:
Add missing import signal

vstinner commented on Modules/signalmodule.c:
```diff
@@ -1607,9 +1607,39 @@ PyErr_CheckSignals(void)
+   The SIGINT signal must be handled by Python, otherwise an exception is
+   raised and return -1. Return 0 on success.
+
+   The GIL doesn't need to be hold to call this function. */
```
You must hold the GIL to call PyErr_SetString().

mcepl replied:
@vstinner That function seems to be called only in [thread_PyThread_interrupt_main](https://github.com/mcepl/cpython/blob/7e414f2f495e73258f55bdef153bc19fb8a9e472/Modules/_threadmodule.c#L1106) and I am really not sure what's the state of GIL there. Of course, the line in the comment should go.

mcepl replied:
Right. The comment is gone.

mcepl committed:
Add forgotten change.

mcepl committed:
It is not true, GIL much be hold for PyErr_SetString() function.

mcepl committed:
Fix the Misc/ACKS with the proper spelling of my name.

vstinner commented on Misc/NEWS.d/next/Library/2016-07-27-11-06-43.bpo-23395.MuCEX9.rst:
```diff
@@ -0,0 +1,3 @@
+``_thread.interrupt_main()`` now raises an exception if the 
+``SIGINT`` signal is ignored or not handled by Python.
+Patch by Viktor Stinner
```
typo: Victor

mcepl replied:
That has been already fixed.

pitrou commented on Doc/c-api/exceptions.rst:
```diff
@@ -494,18 +494,26 @@ Signal Handling
 
-   .. % XXX This was described as obsolete, but is used in
-   .. % _thread.interrupt_main() (used from IDLE), so it's still needed.
+   Deprecated version :c:func:`PyErr_SetInterruptWithErr` which ignores errors.
```
"Deprecated version of ...".

mcepl replied:
@pitrou I don't understand what exactly you object to here. Could you please elaborate?

pitrou replied:
The missing "of" in the sentence above.

mcepl replied:
Right. Fixed.

pitrou commented on Doc/c-api/exceptions.rst:
```diff
@@ -494,18 +494,26 @@ Signal Handling
+   :data:`signal.SIGINT` signal handler will be raised.
+
+   The :data:`signal.SIGINT` signal must be handled by Python, otherwise an
+   exception is raised and return ``-1`` on error. Return ``0`` on success.
```
"... and ``-1`` is returned. ``0`` is returned on success."

pitrou commented on Doc/c-api/exceptions.rst:
```diff
@@ -494,18 +494,26 @@ Signal Handling
+
+   The GIL doesn't need to be hold to call this function.
+
+   .. versionadded:: 3.6
```
3.8, no?

pitrou commented on Lib/test/test_threading.py:
```diff
@@ -1163,12 +1164,51 @@ class BoundedSemaphoreTests(lock_tests.BoundedSemaphoreTests):
+
+class InterruptMainTests(unittest.TestCase):
+    def test_interrupt_main_subthread(self):
+        #Calling start_new_thread with a function that executes interrupt_main
```
Please leave a space after `#` and before the comment.

pitrou commented on Lib/test/test_threading.py:
```diff
@@ -1163,12 +1164,51 @@ class BoundedSemaphoreTests(lock_tests.BoundedSemaphoreTests):
+        t.join()
+
+    def test_interrupt_main_mainthread(self):
+        # Make sure that if interrupt_main is called in main threat that
```
s/threat/thread/

mcepl replied:
That sounds like a very much Freudian typo ;).

pitrou commented on Misc/NEWS.d/next/Library/2016-07-27-11-06-43.bpo-23395.MuCEX9.rst:
```diff
@@ -0,0 +1,3 @@
+``_thread.interrupt_main()`` now raises an exception if the 
+``SIGINT`` signal is ignored or not handled by Python.
+Patch by Viktor Stinner
```
Victor, not Viktor

pitrou commented on Doc/c-api/exceptions.rst:
```diff
@@ -494,18 +494,26 @@ Signal Handling
+   The :data:`signal.SIGINT` signal must be handled by Python, otherwise an
+   exception is raised and return ``-1`` on error. Return ``0`` on success.
+
+   The GIL doesn't need to be hold to call this function.
```
That doesn't sound right. `PyErr_SetString` probably requires the GIL to be held.
If not requiring the GIL is desirable, then errors should be ignored rather than reported.

mcepl replied:
Comment removed.

pitrou commented on Doc/library/_thread.rst:
```diff
@@ -53,8 +53,15 @@ This module defines the following constants and functions:
+   The :data:`signal.SIGINT` signal must be handled by Python, otherwise
+   an exception is raised.
+
+   .. versionchanged:: 3.6
```
3.8

pitrou reviewed (changes requested):
I have a problem with this PR. It doesn't fix `PyErr_SetInterrupt` at all, it requires the caller to switch to another API. I think it would make sense to fix `PyErr_SetInterrupt` so that it does nothing if SIGINT isn't handled by Python.

mcepl committed:
Fix cosmetic issues

mcepl committed:
Fix one more typo.

mcepl committed:
Remove misleading comment about GIL not being needed.

mcepl committed:
Improve documentation about SIGINT handler.

mcepl committed:
Don't introduce new function PyErr_SetInterruptWithErr, fix the old one.

mcepl committed:
PyErr_SetInterrupt could run without GIL, we cannot raise exceptions.

Or return warnings (same problem with memory allocations), or return
errorlevel (changing of API).

pitrou committed:
Some nits
